### PR TITLE
Add `Typing/ProcLiteralReturnTypeRestriction`

### PR DESCRIPTION
**Pull request by nobodywasishere**

### CODE REVIEW

File: src/ameba/rule/typing/proc_literal_return_type_restriction.cr [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  #
+  # ```
+  # task -> { Task.new("execute this command") }
+  # ```
+  #
+  # YAML configuration example:
+  #
+  # ```
+  # Typing/ProcReturnTypeRestriction:
+  #   Enabled: true
```
Comment (veelenga) [inline src/ameba/rule/typing/proc_literal_return_type_restriction.cr:28]: taking into account that all the other suggested rules in Typing scope are disabled, should this one be enabled?

**Review by veelenga — APPROVED** [review]
lgtm

File: src/ameba/rule/typing/proc_literal_return_type_restriction.cr (hunk 1 above)
Comment (nobodywasishere) [reply src/ameba/rule/typing/proc_literal_return_type_restriction.cr:28]: This case IIRC (unlike the others) can lead to footguns when passing procs to/from C, and you already need to define the param types for procs. I can go either way though

Comment (Sija) [reply src/ameba/rule/typing/proc_literal_return_type_restriction.cr:28]: I'd keep it disabled for now. Passing procs to/from C is a) not that common, b) you don't *need* the annotation, so IMO it's more up to the project to enable these.

**Comment by Sija** [comment]
I'm gonna review this (and the rest of `Typing`-related PRs) in the upcoming days, since I have a lot on my plate atm.

**Comment by nobodywasishere** [comment]
No worries, no rush. I thank you for maintaining ameba and taking the time to review my PRs

File: src/ameba/rule/typing/proc_literal_return_type_restriction.cr [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  # Typing/ProcReturnTypeRestriction:
+  #   Enabled: false
+  # ```
+  class ProcLiteralReturnTypeRestriction < Base
+    properties do
+      description "Recommends that proc literals have a return type restriction"
+      enabled false
+    end
+
+    MSG = "Proc literals should have a return type"
```
Comment (Sija) [inline src/ameba/rule/typing/proc_literal_return_type_restriction.cr:36]:
```suggestion
    MSG = "Proc literal should have a return type"
```

File: src/ameba/rule/typing/proc_literal_return_type_restriction.cr [hunk 3]
```diff
@@ -0,0 +1,44 @@
+module Ameba::Rule::Typing
+  # A rule that enforces that Proc literals have a return type.
```
Comment (Sija) [inline src/ameba/rule/typing/proc_literal_return_type_restriction.cr:2]:
```suggestion
  # A rule that enforces that `Proc` literals have a return type.
```

File: src/ameba/rule/typing/proc_literal_return_type_restriction.cr [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  #
+  # For example, these are considered valid:
+  #
+  # ```
+  # my_proc = ->(arg : String) : String { a + "proc" }
```
Comment (Sija) [inline src/ameba/rule/typing/proc_literal_return_type_restriction.cr:7]:
```suggestion
  # greeter = ->(name : String) : String { "Hello #{name}" }
```

File: src/ameba/rule/typing/proc_literal_return_type_restriction.cr [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  # ```
+  #
+  # ```
+  # task -> : Task { Task.new("execute this command") }
+  # ```
+  #
+  # And these are invalid:
+  #
+  # ```
+  # my_proc = ->(arg : String) { a + "proc" }
```
Comment (Sija) [inline src/ameba/rule/typing/proc_literal_return_type_restriction.cr:17]:
```suggestion
  # greeter = ->(name : String) { "Hello #{name}" }
```

File: spec/ameba/rule/typing/proc_literal_return_type_restriction_spec.cr [hunk 6]
```diff
@@ -0,0 +1,20 @@
+require "../../../spec_helper"
+
+module Ameba::Rule::Typing
+  subject = ProcLiteralReturnTypeRestriction.new
+
+  it "passes if a proc literal has a return type" do
+    expect_no_issues subject, <<-CRYSTAL
+      my_proc = ->(var : Nil) : Nil { puts var }
```
Comment (Sija) [inline spec/ameba/rule/typing/proc_literal_return_type_restriction_spec.cr:8]:
```suggestion
      my_proc = ->(var : String) : Nil { puts var }
```

File: spec/ameba/rule/typing/proc_literal_return_type_restriction_spec.cr [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+      my_proc.call(nil)
+      CRYSTAL
+  end
+
+  it "fails if a proc literal doesn't have a return type" do
+    expect_issue subject, <<-CRYSTAL
+      my_proc = ->(var : Nil) { puts var }
```
Comment (Sija) [inline spec/ameba/rule/typing/proc_literal_return_type_restriction_spec.cr:15]:
```suggestion
      my_proc = ->(var : String) { puts var }
```

File: src/ameba/ast/visitors/node_visitor.cr [hunk 8]
```diff
@@ -40,6 +40,7 @@ module Ameba::AST
       Var,
       When,
       While,
+      ProcLiteral,
```
Comment (Sija) [inline src/ameba/ast/visitors/node_visitor.cr:43]: Put it after `NilLiteral`, please.

File: src/ameba/rule/typing/proc_literal_return_type_restriction.cr [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+  #
+  # YAML configuration example:
+  #
+  # ```
+  # Typing/ProcReturnTypeRestriction:
+  #   Enabled: false
+  # ```
+  class ProcLiteralReturnTypeRestriction < Base
+    properties do
+      description "Recommends that proc literals have a return type restriction"
```
Comment (Sija) [inline src/ameba/rule/typing/proc_literal_return_type_restriction.cr:32]:
```suggestion
      description "Disallows Proc literals without return type restrictions"
```

File: src/ameba/rule/typing/proc_literal_return_type_restriction.cr [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+      enabled false
+    end
+
+    MSG = "Proc literal should have a return type"
```
Comment (Sija) [inline src/ameba/rule/typing/proc_literal_return_type_restriction.cr:36]:
```suggestion
    MSG = "Proc literal should have a return type restriction"
```

**Review by Sija — CHANGES_REQUESTED** [review]

File: src/ameba/rule/typing/proc_literal_return_type_restriction.cr [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+  # A rule that enforces that `Proc` literals have a return type.
+  #
+  # For example, these are considered valid:
+  #
+  # ```
+  # greeter = ->(name : String) : String { "Hello #{name}" }
+  # ```
+  #
+  # ```
+  # task -> : Task { Task.new("execute this command") }
```
Comment (veelenga) [inline src/ameba/rule/typing/proc_literal_return_type_restriction.cr:11]:
```suggestion
  # task = -> : Task { Task.new("execute this command") }
```

File: src/ameba/rule/typing/proc_literal_return_type_restriction.cr [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+  # ```
+  #
+  # And these are invalid:
+  #
+  # ```
+  # greeter = ->(name : String) { "Hello #{name}" }
+  # ```
+  #
+  # ```
+  # task -> { Task.new("execute this command") }
```
Comment (veelenga) [inline src/ameba/rule/typing/proc_literal_return_type_restriction.cr:21]:
```suggestion
  # task = -> { Task.new("execute this command") }
```

**Review by veelenga — APPROVED** [review]
lgtm, thank you!

File: spec/ameba/rule/typing/proc_literal_return_type_restriction_spec.cr [hunk 13]
```diff
@@ -0,0 +1,20 @@
+require "../../../spec_helper"
+
+module Ameba::Rule::Typing
+  subject = ProcLiteralReturnTypeRestriction.new
+
+  it "passes if a proc literal has a return type" do
+    expect_no_issues subject, <<-CRYSTAL
+      my_proc = ->(var : String) : Nil { puts var }
+      my_proc.call(nil)
```
Comment (Sija) [inline spec/ameba/rule/typing/proc_literal_return_type_restriction_spec.cr:9]:
```suggestion
```

File: spec/ameba/rule/typing/proc_literal_return_type_restriction_spec.cr [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+      CRYSTAL
+  end
+
+  it "fails if a proc literal doesn't have a return type" do
+    expect_issue subject, <<-CRYSTAL
+      my_proc = ->(var : String) { puts var }
+              # ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ error: Proc literal should have a return type restriction
+      my_proc.call(nil)
```
Comment (Sija) [inline spec/ameba/rule/typing/proc_literal_return_type_restriction_spec.cr:17]:
```suggestion
```

File: spec/ameba/rule/typing/proc_literal_return_type_restriction_spec.cr [hunk 15]
```diff
@@ -0,0 +1,20 @@
+require "../../../spec_helper"
+
+module Ameba::Rule::Typing
+  subject = ProcLiteralReturnTypeRestriction.new
+
+  it "passes if a proc literal has a return type" do
```
Comment (Sija) [inline spec/ameba/rule/typing/proc_literal_return_type_restriction_spec.cr:6]:
```suggestion
  it "passes if a proc literal has a return type restriction" do
```

File: spec/ameba/rule/typing/proc_literal_return_type_restriction_spec.cr [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+    expect_no_issues subject, <<-CRYSTAL
+      my_proc = ->(var : String) : Nil { puts var }
+      my_proc.call(nil)
+      CRYSTAL
+  end
+
+  it "fails if a proc literal doesn't have a return type" do
```
Comment (Sija) [inline spec/ameba/rule/typing/proc_literal_return_type_restriction_spec.cr:13]:
```suggestion
  it "fails if a proc literal doesn't have a return type restriction" do
```

File: src/ameba/rule/typing/proc_literal_return_type_restriction.cr [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+  # ```
+  #
+  # ```
+  # task = -> { Task.new("execute this command") }
+  # ```
+  #
+  # YAML configuration example:
+  #
+  # ```
+  # Typing/ProcReturnTypeRestriction:
```
Comment (Sija) [inline src/ameba/rule/typing/proc_literal_return_type_restriction.cr:27]:
```suggestion
  # Typing/ProcLiteralReturnTypeRestriction:
```

File: src/ameba/rule/typing/proc_literal_return_type_restriction.cr [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+  # ```
+  #
+  # YAML configuration example:
+  #
+  # ```
+  # Typing/ProcLiteralReturnTypeRestriction:
+  #   Enabled: false
+  # ```
+  class ProcLiteralReturnTypeRestriction < Base
+    properties do
```
Comment (Sija) [inline src/ameba/rule/typing/proc_literal_return_type_restriction.cr:31]:
```suggestion
    properties do
      since_version "1.7.0"
```

**Review by Sija — APPROVED** [review]

**Pull request merged**